fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -709,4 +714,183 @@ public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
         expression.getOp().accept(scanCollector, expression, null);
         return scanExprs;
     }
+
+    // get all predicates within and below root
+    public static List<ScalarOperator> getAllPredicates(OptExpression root) {
+        List<ScalarOperator> predicates = Lists.newArrayList();
+        getAllPredicates(root, predicates);
+        return predicates;
+    }
+
+    private static void getAllPredicates(OptExpression root, List<ScalarOperator> predicates) {
+        Operator operator = root.getOp();
+        if (operator.getPredicate() != null) {
+            predicates.add(root.getOp().getPredicate());
+        }
+        if (operator instanceof LogicalJoinOperator) {
+            LogicalJoinOperator joinOperator = (LogicalJoinOperator) operator;
+            if (joinOperator.getOnPredicate() != null) {
+                predicates.add(joinOperator.getOnPredicate());
+            }
+        }
+        for (OptExpression child : root.getInputs()) {
+            getAllPredicates(child, predicates);
+        }
+    }
+
+    // split predicate into three parts: equal columns predicates, range predicates, and residual predicates
+    public static PredicateSplit splitPredicate(ScalarOperator predicate) {
```

Comment:
you should comment what are the equal columns predicates, range predicates, and residual predicates, no every one read the paper.

Reply:
I add it to PredicateSplit